pkg/bitmap/bitmap.go

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright The Dragonfly Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package bitmap
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+	"sync"
+
+	lbm "github.com/openacid/low/bitmap"
+)
+
+const (
+	// sizeOf64BitsLimit limits the max size of array of bitmap.
+	// It is limited by "github.com/openacid/low/bitmap".
+	sizeOf64BitsLimit = (math.MaxInt32 >> 6) + 1
+)
+
+// BitMap is a struct which provides the Get or Set of bits map.
+type BitMap struct {
+	lock        sync.RWMutex
+	bm          []uint64
+	maxBitIndex uint32
+}
+
+// NewBitMap generates a BitMap.
+func NewBitMap(sizeOf64Bits uint32, allSetBit bool) (*BitMap, error) {
+	if sizeOf64Bits > sizeOf64BitsLimit {
+		return nil, fmt.Errorf("sizeOf64Bits %d should be range[0, %d]", sizeOf64Bits, sizeOf64BitsLimit)
+	}
+
+	bm := make([]uint64, sizeOf64Bits)
+	if allSetBit {
+		for i := 0; i < int(sizeOf64Bits); i++ {
+			bm[i] = math.MaxUint64
+		}
+	}
+
+	return &BitMap{
+		bm:          bm,
+		maxBitIndex: (sizeOf64Bits << 6) - 1,
+	}, nil
+}
+
+// RestoreBitMap generate the BitMap by input bytes.
+func RestoreBitMap(data []byte) (*BitMap, error) {
+	if uint64(len(data)) > (sizeOf64BitsLimit << 3) {
+		return nil, fmt.Errorf("sizeOf64Bits %d should be range[0, %d]", uint64(len(data)), sizeOf64BitsLimit<<3)
+	}
+
+	bm := DecodeToUintArray(data)
+
+	return &BitMap{
+		bm:          bm,
+		maxBitIndex: uint32(len(bm)*64 - 1),
+	}, nil
+}
+
+// Get gets the bits in range [start, end]. if set is true, return the set bits.
+// else return the unset bits.
+func (b *BitMap) Get(start uint32, end uint32, setBit bool) ([]*BitsRange, error) {
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	return b.getWithoutLock(start, end, setBit)
+}
+
+// Set sets or cleans the bits in range [start, end]. if setBit is true, set bits. else clean bits.
+func (b *BitMap) Set(start uint32, end uint32, setBit bool) error {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	return b.setWithoutLock(start, end, setBit)
+}
+
+func (b *BitMap) getWithoutLock(start uint32, end uint32, setBit bool) ([]*BitsRange, error) {
```

Comment:
The `get` operation is much slower than the `set` operation. Here is a faster algorithm to implement the `get` operation in most cases and its time complexity is as same as `set`: `O((end-start)>>6)`.

Firstly there is basic function(named `__builtin_ctz` in `C` or `runtime/internal/sys#Ctz64` in `GoLang`) should be introduced: counts trailing (low-order) zeroes.

The main flow is similar to `set` and the core steps to process one `uint64` number `x` are as following:
```golang
if !setBit
    x = ^x
while x:
    // start is the value of the trailing zeros of x
    start = Ctz64(x)
    // remove the trailing zeros of x and counts again
    end = Ctz64(^x>>start) + start
    x = x >> end << end
```

Reply:
Good suggestion! I have  updated it by the algorithm.